airbyte-lib/airbyte_lib/caches/base.py

```diff
@@ -0,0 +1,591 @@
+# Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+
+"""A SQL Cache implementation."""
+
+import abc
+import enum
+from collections.abc import Generator, Iterable
+from contextlib import contextmanager
+from functools import cached_property, lru_cache
+from pathlib import Path
+from typing import TYPE_CHECKING, Any, cast, final
+
+import pandas as pd
+import pyarrow as pa
+import sqlalchemy
+import ulid
+from airbyte_lib.config import CacheConfigBase
+from airbyte_lib.file_writers import FileWriterBase, FileWriterBatchHandle
+from airbyte_lib.processors import BatchHandle, RecordProcessor
+from airbyte_lib.types import SQLTypeConverter
+from airbyte_protocol.models import ConfiguredAirbyteCatalog, ConfiguredAirbyteStream
+from overrides import overrides
+from sqlalchemy import ClauseElement, CursorResult, Executable, MetaData, Table, TextClause, create_engine, text
+from sqlalchemy.engine import Engine
+from sqlalchemy.pool import StaticPool
+
+if TYPE_CHECKING:
+    from sqlalchemy.engine import Connection
+    from sqlalchemy.engine.reflection import Inspector
+
+
+DEBUG_MODE = True
+
+
+class RecordDedupeMode(enum.Enum):
+    APPEND = "append"
+    REPLACE = "replace"
+
+
+class SQLRuntimeError(Exception):
+    """Raised when an SQL operation fails."""
+
+
+class SQLCacheConfigBase(CacheConfigBase):
+    """Same as a regular config except it exposes the 'get_sql_alchemy_url()' method."""
+
+    dedupe_mode = RecordDedupeMode.APPEND
```

Comment:
Worth a discussion, but I expected the default mode to be overwrite which means it will remove the old table and recreate it. Right now with append it will fail on schema changes with no easy way to fix except for manually deleting the table from the destination. What do you think about introducing overwrite here which deletes the target table before loading new data?

Actually, append doesn't feel that important to me in general, especially since we don't have incremental syncs yet.

Reply:
I've updated this significantly. Now, if primary keys exist, a merge insert will be attempted. If primary keys do not exist, then a table swap will be attempted.

We can continue to iterate as we improve our tests, in order to handle all the possible cases.